### ISSUE TITLE
--alwaysStrict not ... always strict 

### ISSUE THREAD

**Issue by jeremyjh** [Microsoft/TypeScript#15041]
**TypeScript Version:**  2.2.2

**Code**

In a React Native project:

```ts
import React from 'react';

let thing = {a: "first"};
Object.freeze(thing);
thing.a = "second";
```
*tsconfig.json*
```json
"compilerOptions":{  
   "target":"es2015",
   "module":"es2015",
   "jsx":"react",
   "outDir":"dist",
   "rootDir":"App",
   "allowSyntheticDefaultImports":true,
   "noImplicitAny":true,
   "strictNullChecks":true,
   "alwaysStrict":true,
   "noUnusedLocals":true,
   "experimentalDecorators":true,
   "preserveConstEnums":true,
   "allowJs":true,
   "moduleResolution":"node",
   "types":[  
      "native-base",
      "redux"
   ],
   "sourceMap":true
},
```
**Expected behavior:**
A run-time exception, because alwaysStrict: true.

**Actual behavior:**
Mutation is silently discarded unless "use strict;" is added to the top of the .ts file.

In #11806 it was identified that "use strict;" is superfluous when module=es2015.  

I am using module=es2015 in a React Native and would like "use strict" to be applied to all of my files, but I'm not sure how to do this now. Setting module to commonjs will bring back "use strict" but breaks RN. Setting target=es5 does not enable "use strict";.

**Comment by aluanhaddad** [comment]
> Setting target=es5 does not enable "use strict";.

This definitely should. It is also completely forwards compatible, valid ES2015 to emit a potentially redundant `"use strict"`.

**Comment by jeremyjh** [comment]
It looks like the real culprit is the babel preset used in React Native's packager. This can be overrode in `.babelrc`. This is why RN is not behaving in spec I think.

**Comment by aluanhaddad** [comment]
@jeremyjh perhaps, but as you say, `tsc` emits non strict code under using your configuration but under ```--target es5```. That is a problem.

**Comment by jeremyjh** [comment]
I'm going to close this; it is not really a Typescript issue, even though Typescript used to support a nice workaround for it. I still actually have not found any workaround other than putting "use strict"; at the top of every module file myself. Building my own .babelrc config with

```json
    ["transform-es2015-modules-commonjs", { strict: true, allowTopLevelThis: true }],
```

Did not resolve the issue for some reason. But I think its more properly a RN issue.

**Comment by aluanhaddad** [comment]
Perhaps link the issue that you file with React Native.

**Comment by jeremyjh** [comment]
@aluanhaddad Actually I found they already had one - facebook/react-native#5316 . Basically they can't fix it, too many libraries out there won't compile under strict and all of your project's node_modules will get processed by the packager. 

I did get an override to work locally in project, and found I also had a couple of dependencies that don't support strict. So its notations at the top of all my files.
